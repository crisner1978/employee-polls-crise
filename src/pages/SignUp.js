import { ExclamationCircleIcon } from "@heroicons/react/24/solid";
import React, { useState } from "react";
import { useForm } from "react-hook-form";
import { checkIfUsernameTaken, createUser } from "../lib/firebase";
import { removeWhiteSpace } from "../lib/removeWhitespace";

export default function SignUp({ user }) {
  const [username, setUsername] = useState(() =>
    removeWhiteSpace(user?.displayName)
  );

  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm({ mode: "onBlur", defaultValues: { username } });

  async function onSubmit(data) {
    const u = `@${removeWhiteSpace(data.username)}`;
    await createUser({ user, username: u });
    window.location.replace('/')
  }

  return (
    <div className="w-full flex flex-col items-center">
      <header className="mt-20 mb-8 text-center">
        <h1 className="font-bold text-xl">Sign Up</h1>
      </header>
      <form onSubmit={handleSubmit(onSubmit)} className="w-80">
        <div className="flex flex-col">
          <label className="font-semibold mb-1">Create username</label>
          <div className="w-full bg-gray-200 py-2 px-4 flex items-center justify-between">
            <input data-testid="input-username"
              className="bg-transparent w-full outline-none"
              {...register("username", {
                required: "Username is required",
                minLength: {
                  value: 5,
                  message: "Must be at least 5 characters",
                },
                maxLength: {
                  value: 20,
                  message: "Must be less than 20 characters",
                },
                validate: (value) => checkIfUsernameTaken(value),
              })}
              type="text"
            />
            {errors.username ? (
              <ExclamationCircleIcon className="h-5 w-5 text-red-400" />
            ) : null}
          </div>
        </div>
        <p data-testid="error-message" className="mt-1 text-red-400">{errors.username?.message}</p>
        <button data-testid="submit-button" className="bg-red-400 hover:brightness-95 transition-all duration-150 ease-out w-full py-2 mt-4 text-white font-semibold text-lg disabled:bg-gray-300">
          Sign Up
        </button>
      </form>
    </div>
  );
}
