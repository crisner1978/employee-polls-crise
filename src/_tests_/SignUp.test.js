import { fireEvent, render, screen } from '@testing-library/react'
import * as React from 'react'
import SignUp from '../pages/SignUp'

describe('SignUp', () => {
  it('will match snapshot', () => {
    var view = render(<SignUp />)
    expect(view).toMatchSnapshot()
  })
})

describe('SignUp Events', () => {
  it('will display an error message if username is less than 5 characters', async() => {
    render(<SignUp />)

    var input = screen.getByTestId('input-username')
    fireEvent.change(input, { target: { value: 'Cris' } })
    var submitButton = screen.getByTestId('submit-button')
    fireEvent.click(submitButton)

    expect(screen.getByTestId('error-message')).toBeInTheDocument()
  })
  it('will display an error message if username is more than 20 characters', () => {
    render(<SignUp />)

    var input = screen.getByTestId('input-username')
    fireEvent.change(input, { target: { value: 'ChristopherWarrenRisner' } })
    var submitButton = screen.getByTestId('submit-button')
    fireEvent.click(submitButton)

    expect(screen.getByTestId('error-message')).toBeInTheDocument()
  })
})
