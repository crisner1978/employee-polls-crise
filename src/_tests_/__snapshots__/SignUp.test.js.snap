// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp will match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="w-full flex flex-col items-center"
      >
        <header
          class="mt-20 mb-8 text-center"
        >
          <h1
            class="font-bold text-xl"
          >
            Sign Up
          </h1>
        </header>
        <form
          class="w-80"
        >
          <div
            class="flex flex-col"
          >
            <label
              class="font-semibold mb-1"
            >
              Create username
            </label>
            <div
              class="w-full bg-gray-200 py-2 px-4 flex items-center justify-between"
            >
              <input
                class="bg-transparent w-full outline-none"
                data-testid="input-username"
                name="username"
                type="text"
              />
            </div>
          </div>
          <p
            class="mt-1 text-red-400"
            data-testid="error-message"
          />
          <button
            class="bg-red-400 hover:brightness-95 transition-all duration-150 ease-out w-full py-2 mt-4 text-white font-semibold text-lg disabled:bg-gray-300"
            data-testid="submit-button"
          >
            Sign Up
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="w-full flex flex-col items-center"
    >
      <header
        class="mt-20 mb-8 text-center"
      >
        <h1
          class="font-bold text-xl"
        >
          Sign Up
        </h1>
      </header>
      <form
        class="w-80"
      >
        <div
          class="flex flex-col"
        >
          <label
            class="font-semibold mb-1"
          >
            Create username
          </label>
          <div
            class="w-full bg-gray-200 py-2 px-4 flex items-center justify-between"
          >
            <input
              class="bg-transparent w-full outline-none"
              data-testid="input-username"
              name="username"
              type="text"
            />
          </div>
        </div>
        <p
          class="mt-1 text-red-400"
          data-testid="error-message"
        />
        <button
          class="bg-red-400 hover:brightness-95 transition-all duration-150 ease-out w-full py-2 mt-4 text-white font-semibold text-lg disabled:bg-gray-300"
          data-testid="submit-button"
        >
          Sign Up
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
